### PR TITLE
Fixes and Update Upgrade test

### PR DESCRIPTION
**Pull request by daniel-butler-irl**
### Description

- getRemoteURL updated to getRemoteOriginURL for more descriptive name
- getSymbolicRef replaced by getCurrentBranch which tries multiple approaches incase one fails on the runtime environment
- CleanTerraformDir added the removes terraform metafiles like state, cache and lock files form the target directory
- DisableTempWorkingDir added to the test options useful for if you need to keep files around after the test when teardown is disabled
- Updated upgrade test to find the upstream URL and origin branch this is to allow a fork to work and have the base branch and repo auto-detected. 
- Option added for `BaseTerraformRepo` and `BaseTerraformBranch` to manually set the base repo and branch for an upgrade test if users are not using a default configuration. These will be overridden by environment variables BASE_TERRAFORM_REPO and BASE_TERRAFORM_BRANCH if set.
- Option added `DisableTempWorkingDir` to disable temporary working directory. Workspace collisions when running in parallel could occur if this is set to true.

### Release required?

- [x] No release. Create manual release v1.21.0 to cleanup the pre-release mess...
- [ ] Patch release (`x.x.X`)
- [ ] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### Merge actions for mergers

- When merging, use a relevant [conventional commit](https://www.conventionalcommits.org/) message that is based on the PR contents and any release notes provided by the PR author. The commit message determines whether a new version of the module is needed, and if so, which semver increment to use (major, minor, or patch).
- Merge by using "Squash and merge".

### CODE REVIEW

File: testhelper/tests.go [hunk 1]
```diff
@@ -231,121 +267,189 @@ func (options *TestOptions) RunTestUpgrade() (*terraform.PlanStruct, error) {
...
+			return nil, fmt.Errorf("failed to copy PR directory to temp: %v", errCopy)
+		} else {
+			logger.Log(options.Testing, "Copied current code to PR branch dir:", prTempDir)
 		}
 
-		if err := remote.Fetch(opts); err != nil {
-			logger.Log(options.Testing, err)
+		baseRepo, baseBranch := common.GetBaseRepoAndBranch(options.BaseTerraformRepo, options.BaseTerraformBranch)
+		if baseBranch == "" || baseRepo == "" {
+			return nil, fmt.Errorf("failed to get default repo and branch: %s %s", baseRepo, baseBranch)
```
Comment (toddgiguere) [inline testhelper/tests.go:347]: Some error conditions in the upgrade function call `testTearDown()` and some do not, should all error returns have a `testTearDown()` as well?

File: testhelper/tests.go [hunk 2]
```diff
@@ -231,121 +267,189 @@ func (options *TestOptions) RunTestUpgrade() (*terraform.PlanStruct, error) {
...
-						logger.Log(options.Testing, "Validating Optional upgrade on Current Branch (PR):", cur.Name())
-						_, resultErr = terraform.ApplyE(options.Testing, options.TerraformOptions)
-						assert.Nilf(options.Testing, resultErr, "Terraform Apply on PR branch has failed")
-					}
-				} else {
-					// if there were issues running InitAndPlan, an Init needs to take place after branch change in order for downstream
-					// terraform to work (like the destroy)
-					terraform.Init(options.Testing, options.TerraformOptions)
-				}
+		assert.Nilf(options.Testing, resultErr, "Terraform Apply on Base branch has failed")
```
Comment (toddgiguere) [inline testhelper/tests.go:400]: If base apply has failed, should we continue? assert() is only going to mark the test as failed but processing will continue, which might not be necessary if base fails.

**Review by toddgiguere — DISMISSED** [review]

File: common/git.go [hunk 3]
```diff
@@ -59,28 +57,133 @@ func (r *realGitOps) gitRootPath(fromPath string) (string, error) {
...
+		log.Printf("Checking if the fromPath: %s directory is the root\n", fromPath)
+		// if fromPath contains .git, then it is the root
+		// otherwise, return the error
+		if _, errNotExist := os.Stat(fromPath + "/.git"); os.IsNotExist(errNotExist) {
+			// check if the current working directory is the root
+			cwd, errGetCwd := os.Getwd()
+			if errGetCwd != nil {
+				return "", fmt.Errorf("failed to determine the Git root path: %s %v", output, err)
+			} else {
+				log.Println("fromPath is not the root")
```
Comment (ocofaigh) [inline common/git.go:70]: you probably meant to print the value of `fromPath` and not the actual string?
But maybe we don't need the log print at all anymore?

File: common/git.go [hunk 4]
```diff
@@ -59,28 +57,133 @@ func (r *realGitOps) gitRootPath(fromPath string) (string, error) {
...
+				log.Printf("Checking if the current working directory: %s is the root\n", cwd)
+				if _, errNotExist := os.Stat(cwd + "/.git"); os.IsNotExist(errNotExist) {
+					log.Println("current working directory is not the root")
+					return "", fmt.Errorf("failed to determine the Git root path: %s %v", output, err)
+				}
+				log.Println("current working directory is the root")
+				return cwd, nil
+			}
+		}
+		log.Println("fromPath is the root")
```
Comment (ocofaigh) [inline common/git.go:80]: same as https://github.com/terraform-ibm-modules/ibmcloud-terratest-wrapper/pull/634/files#r1322737735

File: common/git.go [hunk 5]
```diff
@@ -59,28 +57,133 @@ func (r *realGitOps) gitRootPath(fromPath string) (string, error) {
...
+	cmd := exec.Command("git", "symbolic-ref", "--short", "HEAD")
+	output, _ := cmd.Output()
+	// If the output is empty, try to get the branch name using git rev-parse
+	if string(output) == "" {
+		cmd = exec.Command("git", "rev-parse", "--abbrev-ref", "HEAD")
+		output, _ = cmd.Output()
+	}
+	// If the output is still empty, try to get the branch name using git status
+	if string(output) == "" {
+		cmd := exec.Command("git", "status", "--branch", "--porcelain")
```
Comment (ocofaigh) [inline common/git.go:97]: FYI, this command for me returns as:
```
% git status --branch --porcelain
## catalog-stuff
```

But my branch name is actually:
```
% git symbolic-ref --short HEAD
catalog-stuff
```

Comment (ocofaigh) [reply common/git.go:97]: never mind, I see you parse the branch name later on

File: common/git.go [hunk 6]
```diff
@@ -59,28 +57,133 @@ func (r *realGitOps) gitRootPath(fromPath string) (string, error) {
...
+		re := regexp.MustCompile(`## (.+)\.\.\.`)
+		matches := re.FindStringSubmatch(string(output))
+		if len(matches) != 2 {
+			return "", fmt.Errorf("failed to determine the current branch: unable to parse git status")
+		}
+
+	}
+	branch := strings.TrimSpace(string(output))
+	if branch == "HEAD" {
+		fmt.Println("HEAD means no branch, running in detached mode. This is probable running in GHA")
```
Comment (ocofaigh) [inline common/git.go:113]: probable -> probably

File: common/git.go [hunk 7]
```diff
@@ -59,28 +57,133 @@ func (r *realGitOps) gitRootPath(fromPath string) (string, error) {
...
+	}
+	return branch, nil
+}
+
+func (r *realGitOps) getOriginURL(repoPath string) string {
+	// Determine the URL of the upstream remote (usually "origin")
+	repo := ""
+	cmd := exec.Command("git", "remote", "get-url", "upstream")
```
Comment (ocofaigh) [inline common/git.go:121]: This command only forks for me on a fork. Is that ok?

Non fork:
```
% git remote get-url upstream
error: No such remote 'upstream'
```

Fork:
```
% git remote get-url upstream
git@github.ibm.com:goldeneye/vpc-module
```

File: common/git.go (hunk 3 above)
Comment (daniel-butler-irl) [reply common/git.go:70]: The log message that precedes this is `log.Printf("Checking if the fromPath: %s directory is the root\n", fromPath)` so didn't bother printing again

File: common/git.go (hunk 4 above)
Comment (daniel-butler-irl) [reply common/git.go:80]: same

**Review by ocofaigh — APPROVED** [review]

**Pull request merged**